packages/grid/veilid/server/veilid_streamer.py

```diff
@@ -0,0 +1,393 @@
+
+        # Send STREAM_END request
+        stream_end_message = self.stream_end_struct.pack(
+            RequestType.STREAM_END, call_id
+        )
+        await self._send_request(router, dht_key, stream_end_message)
+
+        if is_request_stream:
+            # This is a new request stream, so we need to wait for
+            # the reply from the receiver
+            logger.debug("Waiting for reply...")
+            response = await buffer_for_holding_reply.message
+            logger.debug("Reply received")
+            # All operations finished, clean up the buffer and return the response
+            del self.buffers[call_id]
+            return response
+        return ResponseType.OK
+
+    async def receive_stream(
+        self,
+        connection: veilid.VeilidAPI,
+        router: veilid.RoutingContext,
+        update: veilid.VeilidUpdate,
+        callback: AsyncReceiveStreamCallback,
+    ) -> None:
+        """Receives a streamed message."""
+        message = update.detail.message
+        prefix = message[:8]
+
+        if prefix == RequestType.STREAM_START:
+            await self._handle_receive_stream_start(connection, update)
+        elif prefix == RequestType.STREAM_CHUNK:
+            await self._handle_receive_stream_chunk(connection, update)
+        elif prefix == RequestType.STREAM_END:
+            await self._handle_receive_stream_end(connection, router, update, callback)
+        else:
+            logger.error(f"[Bad Message] Message with unknown prefix: {prefix}")
+
+    @retry(veilid.VeilidAPIError, tries=4, delay=1, backoff=2)
+    async def _send_request(
+        self, router: veilid.RoutingContext, dht_key: str, request_data: bytes
+    ) -> None:
+        """Send an app call to the Veilid server and return the response."""
+        async with self._send_request_semaphore:
+            response = await router.app_call(dht_key, request_data)
+            if response != ResponseType.OK:
+                raise Exception("Unexpected response from server")
+
+    async def _send_response(
+        self,
+        connection: veilid.VeilidAPI,
+        call_id: veilid.OperationId,
+        response: bytes,
+    ) -> None:
+        """Send a response to an app call."""
+        async with self._send_response_semaphore:
+            await connection.app_call_reply(call_id, response)
+
+    async def _send_ok_response(
+        self, connection: veilid.VeilidAPI, call_id: veilid.OperationId
+    ) -> None:
+        await self._send_response(connection, call_id, ResponseType.OK)
+
+    async def _send_error_response(
+        self, connection: veilid.VeilidAPI, call_id: veilid.OperationId
+    ) -> None:
+        await self._send_response(connection, call_id, ResponseType.ERROR)
+
+    def _calculate_chunks_count(self, message_size: int) -> int:
+        max_chunk_size = self.chunk_size - self.stream_chunk_header_struct.size
+        total_no_of_chunks = message_size // max_chunk_size + 1
```

Comment:
MAYBE: I think this formula could be 
as 
```python3
total_no_of_chunks = math.ceil( message_size / max_chunk_size)
```
as the former formula would fail if the message size is perfectly divisible.

example:

message size = 10
chunk size = 2
total chunks = 5 (the previous formula would return 6)